server.go

```diff
@@ -700,52 +673,75 @@ func (sp *serverPeer) OnGetHeaders(_ *peer.Peer, msg *wire.MsgGetHeaders) {
 	// over with the genesis block if unknown block locators are provided.
 	// This mirrors the behavior in the reference implementation.
 	startIdx := int32(1)
-	for _, hash := range msg.BlockLocatorHashes {
-		height, err := chain.BlockHeightByHash(hash)
+	for _, loc := range locators {
+		height, err := chain.BlockHeightByHash(loc)
 		if err == nil {
 			// Start with the next hash since we know this one.
 			startIdx = height + 1
 			break
 		}
 	}
 
-	// Don't attempt to fetch more than we can put into a single message.
+	// Don't attempt to fetch more than we can put into a single wire
+	// message.
 	if endIdx-startIdx > wire.MaxBlockHeadersPerMsg {
 		endIdx = startIdx + wire.MaxBlockHeadersPerMsg
 	}
 
 	// Fetch the inventory from the block database.
-	hashList, err := chain.HeightRange(startIdx, endIdx)
-	if err != nil {
-		peerLog.Warnf("Header lookup failed: %v", err)
-		return
-	}
-
-	// Generate headers message and send it.
-	headersMsg := wire.NewMsgHeaders()
-	err = sp.server.db.View(func(dbTx database.Tx) error {
-		for i := range hashList {
-			headerBytes, err := dbTx.FetchBlockHeader(&hashList[i])
-			if err != nil {
-				return err
-			}
+	return chain.HeightRange(startIdx, endIdx)
+}
 
-			var header wire.BlockHeader
-			err = header.Deserialize(bytes.NewReader(headerBytes))
+// fetchHeaders fetches and decodes headers from the db for each hash in
+// blockHashes.
+func fetchHeaders(db database.DB, blockHashes []chainhash.Hash) ([]*wire.BlockHeader, error) {
+	headers := make([]*wire.BlockHeader, 0, len(blockHashes))
+	err := db.View(func(dbTx database.Tx) error {
+		rawHeaders, err := dbTx.FetchBlockHeaders(blockHashes)
+		if err != nil {
+			return err
+		}
+		for _, headerBytes := range rawHeaders {
+			h := new(wire.BlockHeader)
+			err = h.Deserialize(bytes.NewReader(headerBytes))
 			if err != nil {
 				return err
 			}
-			headersMsg.AddBlockHeader(&header)
+			headers = append(headers, h)
 		}
-
 		return nil
 	})
+	return headers, err
+}
+
+// OnGetHeaders is invoked when a peer receives a getheaders bitcoin
+// message.
+func (sp *serverPeer) OnGetHeaders(_ *peer.Peer, msg *wire.MsgGetHeaders) {
+	// Ignore getheaders requests if not in sync.
+	if !sp.server.blockManager.IsCurrent() {
+		return
+	}
+
+	blockHashes, err := sp.server.locateBlocks(msg.BlockLocatorHashes,
+		&msg.HashStop)
 	if err != nil {
-		peerLog.Warnf("Failed to build headers: %v", err)
+		peerLog.Errorf("OnGetHeaders: failed to fetch hashes: %v", err)
```

Comment:
I disagree.  Part of running a full node is being able to serve other nodes.  In this case, the input received by the other nodes is not incorrect, and the failure to serve them lies with us rather than them.  In my mind that should be an error and not a warning.  For invalid, unexpected, or deprecated input received from other nodes and and the local administrator (e.g. RPC client), a warning is acceptable iff the program can continue executing.